vaadin-combo-box-flow/src/main/resources/META-INF/resources/frontend/comboBoxConnector.js

```diff
@@ -56,36 +70,44 @@ window.Vaadin.Flow.comboBoxConnector = {
 
       const filterChanged = params.filter !== lastFilter;
       if (filterChanged) {
-        pageCallbacks = {};
         cache = {};
         lastFilter = params.filter;
+        this._debouncer = Debouncer.debounce(
+          this._debouncer,
+          timeOut.after(500),
+          () => {
+            clearPageCallbacks();
+            comboBox.clearCache();
+            if (params.filter === '') {
+              // Fixes the case when the filter changes
+              // from '' to something else and back to ''
+              // within debounce timeout, and the
+              // DataCommunicator thinks it doesn't need to send data
+              comboBox.$server.resetDataCommunicator();
+            }
+          });
+        return;
       }
 
       if (cache[params.page]) {
         // This may happen after skipping pages by scrolling fast
         commitPage(params.page, callback);
       } else {
-        const upperLimit = params.pageSize * (params.page + 1);
-
-        if (filterChanged) {
-          this._debouncer = Debouncer.debounce(
-            this._debouncer,
-            timeOut.after(500),
-            () => {
-              comboBox.$server.setRequestedRange(0, upperLimit, params.filter);
-              if (params.filter === '') {
-                // Fixes the case when the filter changes 
-                // from '' to something else and back to '' 
-                // within debounce timeout, and the
-                // DataCommunicator thinks it doesn't need to send data
-                comboBox.$server.resetDataCommunicator();
-              }
-            });
+        pageCallbacks[params.page] = callback
+        const activePages = Object.keys(pageCallbacks).map(page => parseInt(page));
+        const rangeMin = Math.min(...activePages);
+        const rangeMax = Math.max(...activePages);
+
+        if (rangeMax - rangeMin + 1 !== activePages.length) {
```

Comment:
Let's say the currently active range is from x to y. Only if `params.page` is x - 1 or  y + 1, the range gets extended by the new page. Otherwise the cache is cleared and range gets reset.